networkit/cpp/community/LouvainMapEquation.cpp

```diff
@@ -0,0 +1,583 @@
+}
+
+void LouvainMapEquation::run() {
+    if (hasRun)
+        throw std::runtime_error("Algorithm was already run!");
+
+    Aux::SignalHandler handler;
+
+    Aux::Timer timer;
+    timer.start();
+
+    result.allToSingletons();
+    nextPartition.allToSingletons();
+
+    // remove unused nodes from clustering
+    if (G->numberOfNodes() != G->upperNodeIdBound()) {
+        for (node u = 0; u < G->upperNodeIdBound(); ++u) {
+            if (!G->hasNode(u)) {
+                result.remove(u);
+                if (parallel && parallelizationType == ParallelizationType::Synchronous) {
+                    nextPartition.remove(u);
+                }
+            }
+        }
+    }
+    handler.assureRunning();
+
+    calculateInitialClusterCutAndVolume();
+    timer.stop();
+    DEBUG("init ", timer.elapsedMilliseconds(), " ms");
+
+#ifndef NDEBUG
+    updatePLogPSums();
+#endif
+
+    bool clusteringChanged = false;
+    std::vector<node> nodes = GraphTools::nodeSet(*G);
+
+    count numberOfNodesMoved = 1;
+    for (count iteration = 0; iteration < maxIterations && numberOfNodesMoved > 0; ++iteration) {
+        handler.assureRunning();
+
+        DEBUG("Iteration ", iteration);
+#ifndef NDEBUG
+        DEBUG("Map equation is ", mapEquation());
+#endif
+
+        timer.start();
+        std::shuffle(nodes.begin(), nodes.end(), Aux::Random::getURNG());
+        timer.stop();
+        INFO("shuffle ", timer.elapsedMilliseconds(), " ms");
+
+        timer.start();
+        if (parallelizationType == ParallelizationType::Synchronous) {
+            numberOfNodesMoved = synchronousLocalMoving(nodes, iteration);
+        } else {
+            numberOfNodesMoved = localMoving(nodes, iteration);
+        }
+        timer.stop();
+        INFO("Move iteration ", iteration, " took ", timer.elapsedMilliseconds(), " ms. Moved ",
+             numberOfNodesMoved, " nodes");
+
+        clusteringChanged |= numberOfNodesMoved > 0;
+    }
+
+    handler.assureRunning();
+    if (hierarchical && clusteringChanged) {
+        runHierarchical();
+    }
+    hasRun = true;
+}
+
+count LouvainMapEquation::localMoving(std::vector<node> &nodes, count iteration) {
+    // dummies, since SLM implementation needs more datastructures
+    std::vector<Move> dummyMoves;
+
+    count nodesMoved = 0;
+    if (parallel) {
+#pragma omp parallel
+        {
+            count nm = 0;
+            int tid = omp_get_thread_num();
+            SparseVector<double> &neighborClusterWeights = ets_neighborClusterWeights[tid];
+            if (iteration == 0) {
+                neighborClusterWeights.resize(G->upperNodeIdBound(), 0.0);
+            }
+
+#pragma omp for schedule(guided) nowait
+            for (omp_index i = 0; i < static_cast<omp_index>(nodes.size()); ++i) {
+                if (tryLocalMove(nodes[i], neighborClusterWeights, dummyMoves, false)) {
+                    nm += 1;
+                }
+            }
+
+#pragma omp atomic
+            nodesMoved += nm;
```

Comment:
Here you could use an OpenMP reduction.

Reply:
Yes, but the loop before that is nowait, so there will be little to no contention.